src/rgw/rgw_sal_d4n.h

```diff
@@ -0,0 +1,198 @@
+      c_blk = new cache_block();
+      d4n_cache = new RGWD4NCache();
+    }
+    virtual ~D4NFilterDriver() {
+      delete blk_dir; 
+      delete c_blk;
+      delete d4n_cache;
+    }
+
+    virtual int initialize(CephContext *cct, const DoutPrefixProvider *dpp) override;
+    virtual std::unique_ptr<User> get_user(const rgw_user& u) override;
+
+    virtual std::unique_ptr<Object> get_object(const rgw_obj_key& k) override;
+
+    virtual std::unique_ptr<Writer> get_atomic_writer(const DoutPrefixProvider *dpp,
+				  optional_yield y,
+				  rgw::sal::Object* obj,
+				  const rgw_user& owner,
+				  const rgw_placement_rule *ptail_placement_rule,
+				  uint64_t olh_epoch,
+				  const std::string& unique_tag) override;
+    RGWBlockDirectory* get_block_dir() { return blk_dir; }
+    cache_block* get_cache_block() { return c_blk; }
+    RGWD4NCache* get_d4n_cache() { return d4n_cache; }
+};
+
+class D4NFilterUser : public FilterUser {
+  private:
+    D4NFilterDriver* filter;
+
+  public:
+    D4NFilterUser(std::unique_ptr<User> _next, D4NFilterDriver* _filter) : 
+      FilterUser(std::move(_next)),
+      filter(_filter) {}
+    virtual ~D4NFilterUser() = default;
+
+    virtual int create_bucket(const DoutPrefixProvider* dpp,
+                            const rgw_bucket& b,
+                            const std::string& zonegroup_id,
+                            rgw_placement_rule& placement_rule,
+                            std::string& swift_ver_location,
+                            const RGWQuotaInfo* pquota_info,
+                            const RGWAccessControlPolicy& policy,
+                            Attrs& attrs,
+                            RGWBucketInfo& info,
+                            obj_version& ep_objv,
+                            bool exclusive,
+                            bool obj_lock_enabled,
+                            bool* existed,
+                            req_info& req_info,
+                            std::unique_ptr<Bucket>* bucket,
+                            optional_yield y) override;
+};
+
+class D4NFilterBucket : public FilterBucket {
+  private:
+    D4NFilterDriver* filter;
+
+  public:
+    D4NFilterBucket(std::unique_ptr<Bucket> _next, User* _user, D4NFilterDriver* _filter) :
+      FilterBucket(std::move(_next), _user), 
+      filter(_filter) {}
+    virtual ~D4NFilterBucket() = default;
+   
+    virtual std::unique_ptr<Object> get_object(const rgw_obj_key& key) override;
+};
+
+class D4NFilterObject : public FilterObject {
+  private:
+    D4NFilterDriver* filter;
+
+  public:
+    struct D4NFilterReadOp : FilterReadOp {
+      D4NFilterObject* source;
+
+      D4NFilterReadOp(std::unique_ptr<ReadOp> _next, D4NFilterObject* _source) : FilterReadOp(std::move(_next)),
+										 source(_source) {}
+      virtual ~D4NFilterReadOp() = default;
+
+      virtual int prepare(optional_yield y, const DoutPrefixProvider* dpp) override;
```

Comment:
I will look into whether it'll be an easy addition or not. If it is, it will be part of this PR. Otherwise, I'll introduce it in the next one. I'll update you soon

Reply:
@pritha-srivastava Since we want to merge this PR as soon as possible, I will be implementing the `iterate` overwrite in the next one. For now, there is no way to read the D4N cache data.